ecosystem/sep-0006.md

```diff
@@ -131,7 +185,7 @@ Request Parameters:
 
 Name | Type | Description
 -----|------|------------
-`asset_code` | string | The code of the asset the user wants to deposit with the anchor. The value passed must match one of the codes listed in the [/info](#info) response's deposit object.
+`asset_code` | string | (optional) The code of the asset the user wants to deposit with the anchor. The value passed must match one of the codes listed in the [/info](#info) response's deposit object. This field is mandatory if `quote_id` is not set.
```

Comment:
I think the `asset_code` and `destination_asset` definitions here are a bit confusing. If you're doing a **deposit**, whatever you (as a client/user) are depositing is off-chain, and what you're getting in return from the anchor is on-chain.

There are now two options here:
a) you deposit X and get TokenX (the "traditional" Stellar deposit)
a) you deposit Y and get TokenX (the novel concept introduced by this PR)

The commonality here is that both will lead to you receiving TokenX, which should be one of the tokens specified on the anchor's `/info` endpoint. So it seems to me that what we always meant by `asset_code` is in fact the **destination** asset (TokenX), and what we're introducing now is the concept of a **source asset** when it's not a 1:1 conversion.

So my take is that:
1. `asset_code` should be the destination asset, and it should be one of the assets provided on the `/info` endpoint. This field should **remain mandatory**. This also allows for better backwards API compatibility.
2. we should create a new, optional param called `source_asset` that should have similar characteristics to the ones described for `destination_asset` below and should be accompanied by a `quote_id`.

Reply:
You could also alternatively just make `destination_asset` mandatory, but this would be a breaking API change.

Reply:
Seems like my understanding of the SEP was wrong then. Judging by the existing `asset_code` description:

> `asset_code` | string | The code of the asset the user wants to deposit with the anchor. (...)

I had the understanding that the asset code would be `X` if the user is depositing X to get TokenX. So what you're saying is the asset_code for a SEP-6 deposit would actually be `TokenX`?

Reply:
I don't think your understanding was wrong – my point is that this definition was intentionally ambiguous because effectively `asset_code` meant **both** `X` and `TokenX` as this SEP only described 1:1 asset conversions.

What I'm proposing is that, from now on, we standardize the concept of `asset_code` on both `/withdraw` and `/deposit` flows to always be the **on-chain** asset, and continue being a mandatory parameter on both. In summary:

1. Clients should find all available `asset_code`s on an anchor's `/info` endpoint. If not listed there, the anchor does not support that on-chain asset.
1. For all "traditional" 1:1 deposits or withdrawals, the client only needs to provide `asset_code`.
1. For all SEP38-powered deposits or withdrawals, the client also needs to provide: a `source_asset` (in the case of a deposit); or a `destination_asset` (in the case of a withdrawal).

This should keep the API behavior backwards-compatible for both clients and servers, but also enable them to augment their functionality with SEP-38 FX if they'd like.

Reply:
> 3\. For all SEP38-powered deposits or withdrawals, the client also needs to provide: a `source_asset` (in the case of a deposit); or a `destination_asset` (in the case of a withdrawal).

Do you think `source_asset` and `destination_asset` should be mandatory when using a firm quote with SEP-38 `quote_id`?

Reply:
I think both parameters should be omitted if a firm quote is used, since those assets are already determined when the quote is given via the SEP-38 endpoint.

Reply:
Here are the changes I made based on Alex's suggestions in this thread:
* Make it clear the `asset_code` is the on-chain asset and use `source_asset` for de deposit: https://github.com/stellar/stellar-protocol/pull/973/commits/1844408afb37146f9271c97e19320f3ed9349859
* Make `asset_code` mandatory again: https://github.com/stellar/stellar-protocol/pull/973/commits/f87f49e9f0bdfdbb601e6fec7001042d62b416d6
* Currently, `source_asset` and `destination_asset` are optional and only used for indicative quotes. Firm quotes will use `quote_id` instead.

Reply:
> I think both parameters should be omitted if a firm quote is used, since those assets are already determined when the quote is given via the SEP-38 endpoint.

From an implementations standpoint, I agree with Jake that `asset_code` doesn't need to be mandatory since the anchor should have access to the desired asset through the asset_id.

Reply:
I think conditional optionality of parameters makes APIs confusing, because it introduces many edge cases for the server-side implementer. What if the API caller provides mutually exclusive fields? What if the information in each of these fields is contradictory? If the user sends both a `quote_id` and the other fields (e.g. `source_asset` and `asset_code`), what should take precedence?

It also becomes harder to reason about the implementation from the client perspective: as we build more functionality into the same endpoint, we create more and more conditional rules, to the point that most of our params become optional and we need a whole section in the SEP to explain the possible field combinations. Think of this as the same principle you'd use for writing functions in your code: if you're overloading functions with optional parameters and adding a bunch of `if`s, maybe it's time to break down into several well-scoped functions. APIs are basically remote function calls.

My suggestion is that we take a step back and analyze if what we really should do is to create additional endpoints in this spec, so that each endpoint has a well-defined function and set of mandatory fields, and what we leave as optional is indeed optional information.

Here's what I'm thinking:

**OPTION A (more ergonomic, but also with more changes)**

1. Leave `/deposit` and `/withdraw` as currently defined in SEP-6.
1. Create specialized `/deposit/fx/indicative` and `/withdraw/fx/indicative` endpoints, which take additional `source_asset` and `dest_asset` fields respectively.
1. Create specialized `/deposit/fx/firm` and `/withdraw/fx/firm` endpoints that replace all asset info params with a single, mandatory, `quote_id`.

**OPTION B (less ergonomic, less changes)**

1. Modify the current `/deposit` and `/withdraw` flows to take an additional optional param (`source_asset` or `dest_asset` respectively). `asset_code` is still a mandatory field. The "traditional" and the indicative quote flows will rely on the same endpoint.
1. Create specialized `/deposit/firm-quote` and `/withdraw/firm-quote` endpoints that replace all asset info params with a single, mandatory, `quote_id`.

Reply:
I really like the idea of having dedicated endpoints for each case, so the API gets easier to use and it reduces the chances of mistakes from both the Anchor and Wallet sides. Also, we guarantee that the current SEP-6 behavior will remain unchanged.

> **OPTION A (more ergonomic, but also with more changes)**
> 
> 1. Leave `/deposit` and `/withdraw` as currently defined in SEP-6.
> 2. Create specialized `/deposit/fx/indicative` and `/withdraw/fx/indicative` endpoints, which take additional `source_asset` and `dest_asset` fields respectively.
> 3. Create specialized `/deposit/fx/firm` and `/withdraw/fx/firm` endpoints that replace all asset info params with a single, mandatory, `quote_id`.

About this ☝️ we should additionally update the `/info` data with the new endpoints.

Also, what does `fx` stands for in `/deposit/fx/firm`? Maybe we could use `/deposit/quote/firm` instead?

Reply:
Awesome! FX = foreign exchange. I'm not attached to the endpoint path names though, it was just a way to illustrate my reasoning. I like using `../quote/..` as well.

Reply:
Updated the approach to use dedicated endpoints in https://github.com/stellar/stellar-protocol/pull/973/commits/d1fec88ff9a18166a53b781e231d943db01c14bf. It's best to re-review the whole PR though, rather than spotting the changes in that one commit.

Reply:
I'm going to push back on the idea of new endpoints. I think it introduces other potential issues and increases the development/maintenance burden of the implementer.

What happens if we have another use case that requires the same debate (new endpoint vs. added & omitted params)? We may be able to create new endpoints without issue, but its more likely that this new use case is not mutually exclusive with the use of firm & indicative quotes. We would need to have the following endpoints if we used the same strategy:

- /deposit
- /deposit/quotes/indicative
- /deposit/quotes/firm
- /deposit/new-use-case/A
- /deposit/new-use-case/B
- /deposit/quotes/indicative/new-use-case/A
- /deposit/quotes/indicative/new-use-case/B
- /deposit/quotes/firm/new-use-case/A
- /deposit/quotes/firm/new-use-case/B

I also think the developers of both client and server applications would write functions shared by all endpoints to avoid duplicate code, but it's not unlikely that these shared functions would require conditional logic to handle the variance of each use case. The end result would be code that would look a lot like the function written if there was just one endpoint.

If the implementor duplicates code to avoid this conditional logic, then they increase the surface area for bugs and increase the amount of work needed to make changes that affect all endpoints.

Reply:
> If the implementor duplicates code to avoid this conditional logic, then they increase the surface area for bugs and increase the amount of work needed to make changes that affect all endpoints.

I don't think implementing multiple endpoints means you can't reuse functions in your backend implementation. In fact I think it enforces engineers to think about what is specific to a given endpoint, and what is common logic that can be shared.

> What happens if we have another use case that requires the same debate (new endpoint vs. added & omitted params)? We may be able to create new endpoints without issue, but its more likely that this new use case is not mutually exclusive with the use of firm & indicative quotes. We would need to have the following endpoints if we used the same strategy:
> - /deposit
> - /deposit/quotes/indicative
> - /deposit/quotes/firm
> - /deposit/new-use-case/A
> - /deposit/new-use-case/B
> - /deposit/quotes/indicative/new-use-case/A
> - /deposit/quotes/indicative/new-use-case/B
> - /deposit/quotes/firm/new-use-case/A
> - /deposit/quotes/firm/new-use-case/B

Can you give a concrete example? This is an on/off ramp API, so the basic primitive actions will continue to be **deposit** and **withdraw**. We can have variants for how to accomplish those (like a new alternative method for quotes), but I don't see them layering up.

Also, the alternative to that list of new URLs would be a table that explains all possible combinations of parameters for each desired task. Something like:

* For deposit with new-use-case A, use param X and Y, and do not provide param Z.
* For deposit with new-use-case B, use param X, Z and XX, and do not provide Y
* For deposit with new-use-case-A and indicative quotes, use param X and Y, similarly to the first bullet point, but in the  X use ISO format instead of urlencoded
* For deposit with new-use-case-B using indicative quotes, use param ZZ and nevermind all other params
* ...
* And BTW: all parameters are defined as optional, but they actually aren't.


What is more, what happens if an anchor only wants to implement a subset of these features? If there are multiple endpoints, they can simply not implement the ones they don't support. When we're overloading all possibilities in a single endpoint, it's pretty hard for clients to differentiate whether there's an implementation bug, or if the feature is simply not supported.

It seems pretty clear to me that having multiple endpoints is the way to scale this, while also keeping the SEP understandable. Even if it comes with the (in my opinion, very small) burden of maintaining and integrating with multiple endpoints.

If you're still not convinced, my proposal is that we meet halfway and go with **OPTION B** that I described above. Let me know your thoughts.

Reply:
> Also, the alternative to that list of new URLs would be a table that explains all possible combinations of parameters for each desired task. Something like:
> 
> * For deposit with new-use-case A, use param X and Y, and do not provide param Z.
> * For deposit with new-use-case B, use param X, Z and XX, and do not provide Y
> * For deposit with new-use-case-A and indicative quotes, use param X and Y, similarly to the first bullet point, but in the  X use ISO format instead of urlencoded
> * For deposit with new-use-case-B using indicative quotes, use param ZZ and nevermind all other params
> * ...
> * And BTW: all parameters are defined as optional, but they actually aren't.

This ☝️  sounds very likely to increase complexity in both Wallet and Anchor implementations. If we have `n` mutually exclusive not-really-optional parameters there will be `n!` combinations of parameters that the Wallet and the Anchor will need to figure out. Even though most combinations can be invalid, it'll get pretty hard to figure out which combinations should be supported.

On the other hand, by splitting it into `n` different endpoints and making those parameters mandatory, it'll get much clearer which parameters are needed. The complexity will be `~O(n)`, which should be much preferred than anything exponential like `O(n!)`.

_PS: I'm not so sure if it's `O(n!)` or `O(n^2)`, but my point should remain the same regardless._

Reply:
I'm not able to come up with a concrete use case, and using separate endpoints does make the API contract cleaner.

Reply:
@accordeiro asked me to read through this conversation.

> I think conditional optionality of parameters makes APIs confusing

> If the user sends both a `quote_id` and the other fields (e.g. `source_asset` and `asset_code`), what should take precedence?

I agree with the issue of optionality, but I think the issue of optionality is mostly avoidable without adding additional deposit/withdraw endpoints if we require all fields all the time and validate the inputs against the quote if it is present.

We could also deprecate asset_code and make source_asset and dest_asset required. When asset_code is provided we can make assumptions about source and deposit to satisfy existing use cases. For new implementations this would leave only the quote_id which would be optional, and I don't think that is so bad. In terms of what to do if the assets don't match the quote, we should error, in the same way if the deposit amount doesn't match the quote amount, so I don't think we need to consider precedence.

I also think the path we've chosen to use separate endpoints is fine too, although I have some concerns discussed below.

---

The deposit indicative quote endpoint, does it provide or work off of an indicative quote? I don't see that anywhere. It looks like it will just use whatever the market rate is. Will anyone use it? Consumer products I've used usually provide a firm quote.

If it is using a market rate, I think we should refer to that instead of referring to it as deposit with an indicative quote, because we should point to what is being used for the final conversion.

---

For a firm quote, what happens if you deposit after the expiry? Are you legally required to provide the amount of the firm quote before expired, or does it cancel silently? Does it fallback to market rates? Fwiw I've seen consumer products do the first (legal requirement) and last (market rates) of those things.

---

The use of the terms indicative and firm in the SEP-6 API is a bit confusing. I like that in SEP-38 we refer to these as the price API and the quote API.

Assuming we're still going with separate endpoints, I think they would be easier to understand if they were:

- /deposit – Deposit an asset without conversion.
- /deposit-exchange-with-market-rate – Deposit an asset involving exchange using the market rate.
- /deposit-exchange-with-quote – Deposit an asset involving exchange using a quote.

Reply:
> I agree with the issue of optionality, but I think the issue of optionality is mostly avoidable without adding additional deposit/withdraw endpoints

Let me give it some thought 🤔

> If it is using a market rate, I think we should refer to that instead of referring to it as deposit with an indicative quote, because we should point to what is being used for the final conversion.

I agree this wording is clearer. I'm updating the description in https://github.com/stellar/stellar-protocol/pull/973/commits/4c2769685f1ee625e74a97cf7b0ef18d4816ac8e

> For a firm quote, what happens if you deposit after the expiry? Are you legally required to provide the amount of the firm quote before expired, or does it cancel silently? Does it fallback to market rates? Fwiw I've seen consumer products do the first (legal requirement) and last (market rates) of those things.

I think an anchor should figure out the next steps their own way since they have the user contact. IMO they could either send another quote for approval or return the payment to the user account. I don't think it should fallback to the market price because it can be very different from the user expectations.

Reply:
> > I agree with the issue of optionality, but I think the issue of optionality is mostly avoidable without adding additional deposit/withdraw endpoints
> 
> Let me give it some thought 🤔

One more thing I want to highlight, most of the optionality we're trying to work around exists due to optimizations, not requirements. We're optimizing the request by removing fields we don't really need because the quote has the same data. We don't have to optimize the request, and the overall API complexity may be lower if we don't.

> I don't think it should fallback to the market price because it can be very different from the user expectations.

I agree that the expectations aren't great, but in reality this is what some money transfer services do. For example Wise does this. I think this problem can get complicated because the user has delivered the funds so if they aren't going to go ahead with the exchange how do they get them back to the user?

Reply:
> I agree with the issue of optionality, but I think the issue of optionality is mostly avoidable without adding additional deposit/withdraw endpoints

TBH I'm not a big fan of changing an existent endpoint's interface... my rationale is the endpoints `/deposit` and `/withdraw` where created with the clear purpose of creating a common interface for exchanging equivalent assets that are 1:1, like `USD<>USDC`. Deprecating an API parameter could be very troublesome and bring instability to the network, if an anchor is already implementing it I don't think they should need to make a change just to keep it consistent with the documentation.

On the other hand, I agree that adding 4 new endpoints might be too verbose. Another option that takes into account your suggestions would be to create a single new endpoint called `/exchange` that takes both a `buy_asset` and a `sell_asset` in the forms of `iso4217:USD` and/or `stellar:<asset_code>:<asset_issuer>`. It could also take a `quote_id` for a firm quote. The deposit and withdrawal concepts would kind of lose their purpose in this design because ultimately the user will be just making an exchange where s/he always sends an asset to receive another asset in return.

WDYT @leighmcculloch, @JakeUrban?

Reply:
> I agree that the expectations aren't great, but in reality this is what some money transfer services do. For example Wise does this. I think this problem can get complicated because the user has delivered the funds so if they aren't going to go ahead with the exchange how do they get them back to the user?

Now I get your point. It can get very complicated for the anchor to deal with this case, returning the payment to the user would definitely bring some operational costs. Maybe we could add a statement like this for quoted exchanges: _"If the user deposits the amount to the anchor after the quote expiration, the Anchor has the autonomy to make the asset conversion at market price. The anchor should make this information clear in the `how` field so the Wallet can display it to the user."_

Reply:
> Deprecating an API parameter could be very troublesome and bring instability to the network

That's a good point, this would make a lot of noise for little real advantage to implementers. If we left out the API field change and reused `/deposit` and `/withdraw` for all three use cases we'd end up with two optional parameters:
 - source_asset (for deposit) / destination_asset (for withdrawal)
 - quote_id
Is that right?

I'm not wanting to push that idea. I'm wanting to understand the options.

Reply:
Yes, that's right.

Although keeping only `/deposit` and `/withdraw` and make them use `source_asset` and/or `destination_asset` and/or `quote_id` takes us back to the argument @accordeiro made [previously in this thread](https://github.com/stellar/stellar-protocol/pull/973#discussion_r652073327) when he raised that "the conditional optionality of parameters makes APIs confusing".

Personally, I'd prefer to use a new endpoint for this new feature of converting non-equivalent assets, but remain open to discussions.

Reply:
True, what do you think about combining the two sets of new exchange endpoints into one, so SEP-6 has two sets of APIs: one for simple deposits/withdrawal, and one for exchange deposit/withdrawal?

We can make all the fields required except quote id. Instead of making the RFQAPI responsible for providing the input data for this API for firm quotes, we provide it only as a side-car for linking to a previously locked in rate.

Reply:
> We can make all the fields required except quote id. Instead of making the RFQAPI responsible for providing the input data for this API for firm quotes, we provide it only as a side-car for linking to a previously locked in rate.

Let me know if I understand your suggestions correctly:
* Leave the existing `GET /withdraw` and `GET /deposit` API untouched.
* Create new endpoints for exchanges, say `GET /withdraw-exchange` and `GET /deposit-exchange`
* For the new endpoints, `source_asset` and `destination_asset` are always mandatory and `quote_id` is optional.

Reply:
Yup, that's what I meant. It keeps exchange APIs entirely separate, doesn't affect existing APIs, but reduces the surface area of the total API by 1/3. I also think it is easier to explain: instead of needing to explain 'indicative' and 'firm' names in endpoints the different types of conversion are simply explained as you're converting with or without a quote ahead of time.

Reply:
@leighmcculloch I like this suggestion. It's similar to my `Option B` (2 different sets of endpoints vs. 3), but with better semantics. +1 :)

Reply:
Updated the RFQ strategy in https://github.com/stellar/stellar-protocol/pull/973/commits/ccff6f496e84d8d8f6ab4c340121fdd35764b004.